src/bun.js/node/node_fs_stat_watcher.zig

```diff
@@ -0,0 +1,508 @@
+
+const PathWatcher = @import("./path_watcher.zig");
+
+const VirtualMachine = JSC.VirtualMachine;
+const EventLoop = JSC.EventLoop;
+const PathLike = JSC.Node.PathLike;
+const ArgumentsSlice = JSC.Node.ArgumentsSlice;
+const Output = bun.Output;
+const string = bun.string;
+const StoredFileDescriptorType = bun.StoredFileDescriptorType;
+const Environment = bun.Environment;
+
+const StatsSmall = bun.JSC.Node.StatsSmall;
+const StatsBig = bun.JSC.Node.StatsBig;
+
+const log = bun.Output.scoped(.StatWatcher, false);
+
+fn statToJSStats(globalThis: *JSC.JSGlobalObject, stats: bun.Stat, bigint: bool) JSC.JSValue {
+    if (bigint) {
+        return StatsBig.initWithAllocator(globalThis.allocator(), stats).toJS(globalThis);
+    } else {
+        return StatsSmall.initWithAllocator(globalThis.allocator(), stats).toJS(globalThis);
+    }
+}
+
+/// This is a singleton struct that contains the timer used to schedule restat calls.
+pub const StatWatcherScheduler = struct {
+    timer: ?*uws.Timer = null,
+
+    head: ?*StatWatcher = null,
+    tail: ?*StatWatcher = null,
+
+    task: JSC.WorkPoolTask = .{ .callback = &workPoolCallback },
+
+    mutex: Mutex = Mutex.init(),
+
+    pub fn init(allocator: std.mem.Allocator, _: *bun.JSC.VirtualMachine) *StatWatcherScheduler {
+        var this = allocator.create(StatWatcherScheduler) catch @panic("out of memory");
+        this.* = .{};
+        return this;
+    }
+
+    pub fn append(this: *StatWatcherScheduler, watcher: *StatWatcher) void {
+        log("append new watcher {s}", .{watcher.path});
+        std.debug.assert(watcher.closed == false);
+        std.debug.assert(watcher.next == null);
+
+        this.mutex.lock();
+        defer this.mutex.unlock();
+
+        if (this.head == null) {
+            this.head = watcher;
+            this.tail = watcher;
+
+            watcher.last_check = std.time.Instant.now() catch unreachable;
+
+            const vm = watcher.globalThis.bunVM();
+            this.timer = uws.Timer.create(
+                vm.event_loop_handle orelse @panic("UWS Loop was not initialized yet."),
+                this,
+            );
+
+            this.timer.?.set(this, timerCallback, watcher.interval, 0);
+            log("I will wait {d} milli initially", .{watcher.interval});
+        } else {
+            this.tail.?.next = watcher;
+            this.tail = watcher;
+            this.timer.?.set(this, timerCallback, 1, 0);
+        }
+    }
+
+    pub fn timerCallback(timer: *uws.Timer) callconv(.C) void {
+        var this = timer.ext(StatWatcherScheduler).?;
+        JSC.WorkPool.schedule(&this.task);
+    }
+
+    pub fn workPoolCallback(task: *JSC.WorkPoolTask) void {
+        var this: *StatWatcherScheduler = @fieldParentPtr(StatWatcherScheduler, "task", task);
+        // Instant.now will not fail on our target platforms.
+        const now = std.time.Instant.now() catch unreachable;
+
+        var prev: *StatWatcher = this.head.?;
```

Comment:
since you used a linked list, this can be made race-free by swapping this.head with null

and at the end of the loop, swapping it back and appending the new head to the tail